src/PIL/Image.py

```diff
@@ -3069,7 +3069,15 @@ def frombuffer(mode, size, data, decoder_name="raw", *args):
     return frombytes(mode, size, data, decoder_name, args)
 
 
-def fromarray(obj, mode=None):
+class SupportsArrayInterface(Protocol):
+    """
+    An object that has an ``__array_interface__`` dictionary.
+    """
+
+    __array_interface__: dict[str, Any]
```

Comment:
We should probably use a `@property` like [numpy](https://github.com/numpy/numpy/blob/e59c074842e3f73483afa5ddef031e856b9fd313/numpy/__init__.pyi#L991):

```suggestion
    @property
    def __array_interface__(self) -> dict[str, Any]:
        raise NotImplementedError()
```

As it is written now, it requires read-write access, whearas we only need read access.

See https://mypy-play.net/?mypy=latest&python=3.12&gist=b20b51c7cf6c3360a24d4cf77fb1ca64:

> main.py:22: error: Incompatible types in assignment (expression has type "TestProperty", variable has type "SupportsArrayInterfaceAttribute")  [assignment]
> main.py:22: note: Protocol member SupportsArrayInterfaceAttribute.__array_interface__ expected settable variable, got read-only attribute